syntax = "proto3";
package woops.rpc;

service PsService {
    rpc CheckAlive(CheckAliveRequest) returns (CheckAliveResponse) {}
    rpc Assign(AssignRequest) returns (AssignResponse) {}
    rpc Update(stream UpdateRequest) returns (stream UpdateResponse) {}
    rpc Pull(stream PullRequest) returns (stream PullResponse) {}
}

message CheckAliveRequest {

}

message CheckAliveResponse {
    bool status = 1; 
}

message AssignRequest {
    int32 client = 1;
    string name = 2;
    bytes param = 3;
}

message AssignResponse {
    bytes param = 1;
}

message UpdateRequest {
    string name = 1;
    int32 iteration = 2;
    bytes delta = 3;
}

message UpdateResponse {
}

message PullRequest {
    string name = 1;
    int32 iteration = 2;
}

message PullResponse {
    string name = 1;
    int32 iteration = 2;
    bytes param = 3;
}
